src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -331,20 +330,134 @@ class SchemaService(private val options: Neo4jOptions, private val driverCache:
-    queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+    expectedQueryTypes.size == 0 || expectedQueryTypes.contains(queryType)
+  } catch {
+    case e: Throwable => {
+      log.error("Query not compiled because of the following exception:", e)
+      false
+    }
   }
 
   def isValidQueryCount(query: String): Boolean = {
-    val resultSummary = session.run(s"EXPLAIN $query").consume()
-    val queryType = resultSummary.queryType()
-    val plan = resultSummary.plan()
-    val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
-    val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
-    isReadOnly && hasCountIdentifier
+    try {
+      val resultSummary = session.run(s"EXPLAIN $query").consume()
+      val queryType = resultSummary.queryType()
+      val plan = resultSummary.plan()
+      val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+      val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
+      isReadOnly && hasCountIdentifier
+    } catch {
+      case e: Throwable => {
+        log.error("Query not compiled because of the following exception:", e)
+        false
+      }
+    }
   }
 
+  private def createIndexOrConstraint(action: OptimizationType.Value, label: String, props: Seq[String]) = {
+    if (action == OptimizationType.NONE) {
+      log.info("No optimization type provided")
+      Unit
+    } else {
+      val quotedLabel = label.quote()
+      val quotedProps = props.map(prop => s"${Neo4jUtil.NODE_ALIAS}.${prop.quote()}").mkString(", ")
+      val (querySuffix, creationType, uniqueness) = action match {
+        case OptimizationType.INDEX => (s"FOR (${Neo4jUtil.NODE_ALIAS}:$quotedLabel) ON ($quotedProps)", "INDEX", "NONUNIQUE")
+        case OptimizationType.NODE_CONSTRAINTS => {
+          val assertType = if (props.size > 1) "NODE KEY" else "UNIQUE"
+          (s"ON (${Neo4jUtil.NODE_ALIAS}:$quotedLabel) ASSERT ($quotedProps) IS $assertType", "CONSTRAINT", "UNIQUE")
+        }
+      }
+      val actionName = s"spark_${creationType}_$label".quote()
+      val queryPrefix = s"CREATE $creationType $actionName"
+
+      val queryCheck = """CALL db.indexes() YIELD labelsOrTypes, properties, uniqueness
+          |WHERE labelsOrTypes = $labels AND properties = $properties AND uniqueness = $uniqueness
+          |RETURN count(*) > 0 AS isPresent""".stripMargin
+
+      val isPresent = session.run(queryCheck, Map("labels" -> Seq(label).asJava,
+          "properties" -> props.asJava,
+          "uniqueness" -> uniqueness).asJava)
+        .single()
+        .get("isPresent")
+        .asBoolean()
+
+      val status = if (isPresent) {
+        "KEPT"
+      } else {
+        val query = s"$queryPrefix $querySuffix"
```

Comment:
this chunk would be better refactored into a reusable function

- Log a message
- Run a query
- Log a failure if it happens

(this gets done in several places)

Reply:
is it ok if we do it in a separate PR?

Reply:
yes that's fine